pkg/credentialprovider/apis/criauth/v1alpha1/types.go

```diff
@@ -0,0 +1,118 @@
+// each plugin as specified by the CRIAuthPlugin type.
+type CRIAuthPluginConfig struct {
+	metav1.TypeMeta `json:",inline"`
+
+	// Plugins is a list of CRI auth plugins that will be enabled by the kubelet
+	Plugins []CRIAuthPlugin `json:"plugins"`
+}
+
+// CRIAuthPlugin is used by the kubelet container runtime to match the
+// image property string (from the container spec) with exec-based credential provider
+// plugins that provide container registry credentials.
+type CRIAuthPlugin struct {
+	// name is the name of the credential plugin.  It must match the name of the
+	// executable as seen by the kubelet.
+	Name string `json:"name"`
+
+	// imageMatchRefs contains information about the matching image URLs this plugin should
+	// handle along with the cache duration of credentials for any matching images.
+	// Credentials are cached independently per imageMatchRef.
+	ImageMatchRefs []ImageMatchRef `json:"imageMatchRefs"`
+
+	// ExtraArgs specifies arguments to be passed to the plugin via environment variables.
+	// +optional
+	ExtraArgs []PluginArg `json:"extraArgs,omitempty"`
+}
+
+// ImageMatchRef holds information for what image URLs a plugin should handle.
+// The plugin will be called by a kubelet if an image requiring credentials matches
+// the information in imageMatchRef. It also indicates how long the auth plugin
+// should cache credentials in-memory.
+type ImageMatchRef struct {
+	// matchURL is a list of strings used to match against the image property
+	// (sometimes called "registry path") to determine which images to provide
+	// credentials for.  If one of the strings matches the image property, then the
+	// CRIAuthPlugin will be used by kubelet to provide credentials
+	// for the image pull.
+	//
+	// The image property of a container supports the same syntax as the docker
+	// command does, including private registries and tags. A registry path is
+	// similar to a URL, but does not contain a protocol specifier (https://).
+	//
+	// Each matchURL string is a pattern which can optionally contain
+	// a port and a path, similar to the image spec.  Globs can be used in the
+	// hostname (but not the port or the path).
+	//
+	// Globs are supported as subdomains (*.k8s.io) or (k8s.*.io), and
+	// top-level-domains (k8s.*).  Matching partial subdomains is also supported
+	// (app*.k8s.io).  Each glob can only match a single subdomain segment, so
+	// *.io does not match *.k8s.io.
+	//
+	// The image property matches when it has the same number of parts as the
+	// ImageMatcher string, and each part matches.  Additionally the path of
+	// ImageMatcher must be a prefix of the target URL. If the ImageMatcher
+	// contains a port, then the port must match as well.
+	MatchURL string `json:"matchUrl"`
+
+	// cacheDuration is the duration the plugin will cache credentials in-memory
+	// for images that matches this reference as indicated by the matchURL field.
+	CacheDuration metav1.Duration `json:"cacheDuration"`
```

Comment:
azure had some special rules around caching empty results... is this expressive enough to externalize the in-tree azure behavior?

Reply:
> azure had some special rules around caching empty results... is this expressive enough to externalize the in-tree azure behavior?

I think so, today the only check in Azure for not caching is if the provider returned no results:

https://github.com/kubernetes/kubernetes/blob/1420b377e40cc653819916c7db2f4a6ac9cf453b/pkg/credentialprovider/azure/azure_credentials.go#L63

In the plugin case, caching behavior is determined by the cache duration in the plugin response and caching never happens on error, so there isn't a case where the kubelet caches an empty DockerConfig from the plugin.